### PR TITLE
Copy meta attributes to shared definitions when creating a `$ref`

### PR DESCRIPTION
**Pull request by jloleysens**
I would like to update a specific piece of behaviour.  Given the following Joi schema description:

```json
{
  "type": "any",
  "flags": {
    "id": "person"
  },
  "metas": [
    {
      "x-expandable": true,
      "x-lookup": "name",
      "x-supported-lang": ["zh-CN", "en-US"],
      "deprecated": true,
      "readOnly": false,
      "writeOnly": true,
      "z-ignore": true
    }
  ]
}

```

We can get the following open api flavoured JSON Schema (note: `deprecated` and other `x-` attrs are copied to the `schemas` collection):

```json
{
  "$ref": "#/components/schemas/person",
  "schemas": {
    "person": {
      "nullable": true,
      "anyOf": [
        {
          "items": {},
          "type": "array"
        },
        {
          "type": "boolean"
        },
        {
          "type": "number"
        },
        {
          "type": "object"
        },
        {
          "type": "string"
        }
      ],
      "deprecated": true,
      "x-expandable": true,
      "x-lookup": "name",
      "x-supported-lang": ["zh-CN", "en-US"],
      "readOnly": false,
      "writeOnly": true
    }
  }
}
```

Current behaviour is that these attrs are "left behind" on the `$ref` object like:

```json
{
  "$ref": "#/components/schemas/person",
  "deprecated": true,
  "x-expandable": true,
  "x-lookup": "name",
  "x-supported-lang": ["zh-CN", "en-US"],
  "readOnly": false,
  "writeOnly": true
  "schemas": {
    "person": {
      "nullable": true,
      "anyOf": [
        {
          "items": {},
          "type": "array"
        },
        {
          "type": "boolean"
        },
        {
          "type": "number"
        },
        {
          "type": "object"
        },
        {
          "type": "string"
        }
      ],
    }
  }
}
```

### CODE REVIEW

File: lib/parsers/json.js [hunk 1]
```diff
@@ -775,6 +772,13 @@ class JoiJsonSchemaParser {
       delete schema.type
     }
   }
+
+  _moveSchemaToDefinitions(schema, joiSpec, definitions, schemaId) {
+    definitions[schemaId] = schema
+    return {
+      $ref: `${this._getLocalSchemaBasePath()}/${schemaId}`
+    }
+  }
```
Comment (jloleysens) [inline lib/parsers/json.js:781]: Preserve current top-level behaviour

File: lib/parsers/open-api.js [hunk 2]
```diff
@@ -126,6 +120,26 @@ class JoiOpenApiSchemaParser extends JoiJsonSchemaParser {
...
+    if (!_.isEmpty(joiSpec.metas)) {
+      _.each(joiSpec.metas, meta => {
+        _.each(meta, (value, key) => {
+          if (key.startsWith('x-') || key === 'deprecated' || key === 'readOnly' || key === 'writeOnly') {
+            schema[key] = value
+          }
+        })
+      })
+    }
+  }
```
Comment (jloleysens) [inline lib/parsers/open-api.js:134]: Do we want to do this for any other of the targets?

File: lib/parsers/json.js (hunk 1 above)
Comment (kenspirit) [reply lib/parsers/json.js:781]: I think it might not need to preserve the top-level behavior when there is "id" for the schema as it can be referenced anyway.

File: lib/parsers/open-api.js (hunk 2 above)
Comment (kenspirit) [reply lib/parsers/open-api.js:134]: Other target should be appropriate as well, thanks.

File: lib/parsers/json.js (hunk 1 above)
Comment (jloleysens) [reply lib/parsers/json.js:781]: Right, what I meant was that I'll just leave this base json converter unchanged since it also has certain copy behaviour and does not seem to support any metas. Does that make sense or did I miss something?

File: lib/parsers/open-api.js (hunk 2 above)
Comment (jloleysens) [reply lib/parsers/open-api.js:134]: Done in 1f8f068

**Comment by jloleysens** [comment]
Thanks for the review @kenspirit ! Let me know what you think, I believe I have addressed your feedback.

File: lib/parsers/open-api-3.1.js [hunk 3]
```diff
@@ -32,6 +26,26 @@ class JoiOpenApiSchemaParser extends JoiJsonSchemaParser {
...
+    }
+  }
+
+  _moveSchemaToDefinitions(schema, joiSpec, definitions, schemaId) {
+    this._copyMetasToSchema(joiSpec, schema)
+    definitions[schemaId] = schema
+    return {
+      $ref: `${this._getLocalSchemaBasePath()}/${schemaId}`
+    }
+  }
```
Comment (kenspirit) [inline lib/parsers/open-api-3.1.js:48]: I wonder if `_copyMetasToSchema` and `_moveSchemaToDefinitions` can be placed at `parsers/json.js` to standardize behavior.  Then we can have an overridable function `_isMetaQualifiedToCopy` or something  similar, then each parser can return true/false depending on the key checking logic, and then it do the copy action accordingly.

File: lib/parsers/json.js [hunk 4]
```diff
@@ -75,9 +75,9 @@ const LOGICAL_OP_PARSER = {
  */
 function isJoiOverride(e) {
   return typeof e === 'object'
-         && e !== null
-         && Object.keys(e).length === 1
-         && e.override === true
+    && e !== null
+    && Object.keys(e).length === 1
+    && e.override === true
```
Comment (jloleysens) [inline lib/parsers/json.js:80]: Unrelated formatting updates

File: lib/parsers/open-api.js [hunk 5]
```diff
@@ -1,6 +1,48 @@
 const _ = require('lodash')
 const JoiJsonSchemaParser = require('./json-draft-04')
 
+function isKnownMetaKey(key) {
+  return key.startsWith('x-') || key === 'deprecated' || key === 'readOnly' || key === 'writeOnly'
+}
+
+function isKnownKey(key) {
```
Comment (jloleysens) [inline lib/parsers/open-api.js:8]: Tried to stick as close to previous `_.pick` implementation as possible using `Set` for more optimised `.has` now that this parser is using `_.pickBy`

**Review by jloleysens — COMMENTED** [review]
I've reduced code repetition, let me know what you think!

File: lib/parsers/json.js [hunk 6]
```diff
@@ -775,6 +777,10 @@ class JoiJsonSchemaParser {
       delete schema.type
     }
   }
+
+  // _copyMetasToSchema is intended to be overridden by child parsers.
+  // eslint-disable-next-line no-unused-vars
+  _copyMetasToSchema(joiSpec, schema) { }
```
Comment (kenspirit) [inline lib/parsers/json.js:783]: What do you think if the `_copyMetasToSchema` implementation is here, and the `isKnownMetaKey` is the function required to be overridden by child parsers? 😃

Comment (jloleysens) [reply lib/parsers/json.js:783]: IMO this way provides more flexibility to child parsers if their needs on how meta is added ever diverges, although it creates a bit more code repetition. I'll add your suggested approach though.

Comment (jloleysens) [reply lib/parsers/json.js]: Done in 7b44c60
